### ISSUE TITLE
A new protocol

### ISSUE THREAD

**Issue by chris-belcher** [JoinMarket-Org/joinmarket#171]
Issue which combines all the features we want when the protocol is updated.

Talking to some people, it seems doing a breaking change, a "hardfork", is quite an unpopular idea. It would split the liquidity, updating would end up having to be rushed by some people, giving not enough time to check the code. So it seems the compromise solution is to code it that new bots will accept both protocols and the old protocol is gradually phased out.

Stuff to fix:
#88 Fix bug where maker must known the private key of the coinjoin address.
#83 Using p2sh / multisig as inputs.
#45 Some cleanup, changing names, adding a field
#29 Minimum other maker count field
https://github.com/chris-belcher/joinmarket/pull/90#issuecomment-108648823 Allowing taker to not authenticate
#120 Fixing miner fees. Changing maker's tx fee contribution to a fee/kB and having a new field of "lowest fee/kB tx this maker is willing to sign off for"

It should also be noted that the patientsendpayment.py script *does not work at all* in the current protocol, because makers must know the private key of the coinjoin address they're sending to.

#90 Some code already written
The branch https://github.com/chris-belcher/joinmarket/tree/newprotocol

**Comment by chris-belcher** [comment]
More code / bugfix to add #170

**Comment by chris-belcher** [comment]
Allow option of maker not having a change address https://github.com/chris-belcher/joinmarket/issues/27#issuecomment-74863580

**Comment by dcousens** [comment]
@chris-belcher is the algorithm/protocol written up anywhere?

**Comment by chris-belcher** [comment]
Not really

**Comment by kristovatlas** [comment]
I might try to help with documentation in the future...

**Comment by chris-belcher** [comment]
Worth noting that this won't be the last time we update the protocol. We'll probably end up adding segwit and op_schnorr to JoinMarket eventually too. So the code for new protocol should be written in a way that makes adding more order types easier.

**Comment by chris-belcher** [comment]
And anything we decide on regarding this issue will be included in a new protocol too: #156

**Comment by AdamISZ** [comment]
#568  is now a protocol change (so as to find a logical way to use multiple messaging servers without impersonation).

I am not sure what people feel comfortable with as a reasonable set of changes; anti- #156 is priority 1, it seems; in line with #343 goals for 0.2 I think the above multi-mc should also be included. Re: the above list, I'm not sure, but maybe we should hash out, here, what set is both highly desired, and also a realistic goal for a new version.

**Comment by chris-belcher** [comment]
I agree with that, I was thinking of including a few small other things in the first new protocol if theres time/effort (like fixing patientsendpayment.py and the #537 signing change)

**Comment by chris-belcher** [comment]
Yes that's a good idea.

One thing we should get from this protocol break is that the codebase is written so that it's easier to add future new offer types (coinswap, segwit, #229)

**Comment by AdamISZ** [comment]
OK, got started. Tweaking/testing a basic podle PR on the now-existing 0.2.0 branch. Won't be final but will be something to look at.

Re: new ordertypes, yeah, good thing is I think it mostly works out of the box (bots just ignore what's not in their list of known ordertypes right).

**Comment by chris-belcher** [comment]
That's what I meant regarding the hardcoded ordertypes. And both relorder and absorder today invoke the same functions when the order is filled (CoinJoinOrder in maker.py.) When adding segwit and coinswap it would need an entirely different function.
I was thinking an OOP way of organising this. There's an abstract class which has functions called things like handle_offer(), subclasses can override it and implement segwit or whatever it might be.

Plus a nitpick: rename it all to offer instead of order.

**Comment by AdamISZ** [comment]
https://gist.github.com/AdamISZ/baf93ce2589854a7992383b3c69fae13

An overview, so we can see what the new message protocol looks like. Of course, only a first cut.

I'll just push the initial PoDLE update to the 0.2.0 branch now, after all we are just going to keep working on it, so I guess no point having some kind of PR process really.

**Comment by AdamISZ** [comment]
43f6b2a0cd0467312ffe3ec99090f659d9fa2221 is specifically to address the formatting inconsistencies of `blockchaininterface.py` and `support.py`, it does nothing else. It's a little painful, but you might want to pass an eye over it to make sure nothing insane happens. The code passes regtest.

It seemed wise to isolate it in that one commit.

**Comment by AdamISZ** [comment]
5b0e35641196d56cf3674927d06e117cfbf89e0f addresses #88 after prompting by @chris-belcher on IRC.

This implementation simply takes the first utxo that the maker has decided to spend and extracts its key from the wallet and uses that for the authorisation pubkey and signature.

Hence the protocol is changed in line with the proposal in #88:
from

```
M*: !ioauth ulist cjpub changeA B(mencpubkey) (NS)
```
to
```
M*: !ioauth ulist auth_pub cj_addr changeA B(mencpubkey) (NS)
```

and the taker verifies that the address derived from `auth_pub` appears in one of the input utxos.

Presumably other cases could be handled with additional code, but without needing to change the protocol again.

I'll update the protocol gist to reflect this change.

**Comment by chris-belcher** [comment]
I've looked at 43f6b2a and it looks fine. 7d83041 is very good. Thanks for doing it.

**Comment by AdamISZ** [comment]
Another thing I want to mention before I forget: the current PoDLE implementation requires secp256k1 since it makes extensive use of EC point arithmetic. I'm leaning more towards ditching non-secp256k1 (it also finally clears out the mess from the bitcoin module).

**Comment by AdamISZ** [comment]
As per IRC, 8ecae434c24fdec1eab42d6e8e0b4a59518b365e for standardised bitcoin signature format with prefix.

Whole build test passes except tx_broadcast. I think I know what that is, but can be addressed later.

**Comment by AdamISZ** [comment]
@chris-belcher thanks for paying attention, fixed in ad1d8768c0f9b251437266fc1926f0044f73f285 ; it is not actually a change, just a tidy up, as you said pybitcointools version was already "standard".

However I did get a chance to investigate more how it works in libsecp256k1 + Core; they use SignCompact with a recoverable signature, which is a recid single byte + 64 bytes. I'm left unsure about the whether, to support hardware wallet, we should be using this kind of signature; I guess probably so. Afaict this should not be a problem as the python binding supports it. But, it's slightly non-trivial work.

**Comment by AdamISZ** [comment]
I suggested @adlai might just directly commit #423 to 0.2.0 ; it is sort of squishily incompatible with old bots; and see https://github.com/JoinMarket-Org/joinmarket/pull/423#issuecomment-232646767

**Comment by AdamISZ** [comment]
I made some improvements to the podle/commitment code today and added some tests, but mostly I was teasing out the details of how to use it such that it's practical. Once it seems in a sensible state I'll push again.

[This](https://gist.github.com/AdamISZ/baf93ce2589854a7992383b3c69fae13/revisions#diff-3de9d4b4dba28a5e1f82c3d956ab363d) update to the gist is mostly clarification but please cast your eyes.

Also, are we still of the opinion that it may be sensible to remove the MITM protection element from the taker side? I was never 100% sure but this is the time to make the decision, if we are going to, of course.

**Comment by chris-belcher** [comment]
I believe we originally said make it optional. But if it's okay to skip it some of the time then I guess it's okay to skip it all the time(?)

**Comment by AdamISZ** [comment]
Yeah, not sure. Sorry for delay. I'm deep into re-working the commitment stuff so it's flexible enough to use in practice. Will write everything down shortly.

**Comment by AdamISZ** [comment]
0568b5ccbf41556ea1d572f95a42e02825af7d6b is a fairly big re-work of the commitments code. Here's the details:

In `!fill`, there is a `commit_type_byte` prefixing the commitment itself ("P" for PoDLE now). So later we could add different commitment types.

The opening of the commitment, in `!auth` is now one field in the message, with its own serialization rule (here just a separator `|` within the field), again to enable flexibility in future; any style of commitment will have its own serialization of the opening here.

The used commitments by the taker are stored in `commitments.json`, in the first key `used` in the dict. The other key is `external`, in which a taker/user can store a list of commitments generated from external wallets/sources: the format is shown [here](https://github.com/JoinMarket-Org/joinmarket/commit/0568b5ccbf41556ea1d572f95a42e02825af7d6b#diff-67f3c1e1197f4928c39bc5262345f35dR310). 

The `podle.py` module is refactored a bit for more flexibility; there is `podle.PoDLE` class which enables creation either with the privkey and the utxo, or with the public keys and signature (i.e. creation without knowing the private key, to allow verification of other's opened commitments). This module also has a higher level `generate_podle` from a given list of utxos and privkeys; it will either generate the PoDLE from that list, or if that fails, it will try to source the PoDLE commitment from the `external` commitment data in the `commitments.json` file as explained above.

The current amendments to the `CoinJoinTx` code will provide the list of utxos that are inputs to the transaction, so it'll either need to be the case that they have not all been used `taker_utxo_retries` times (where that's set by each user individually, but a default of perhaps 3 to start), or the provided external commitments have not been used the same number of times, in order for the !auth message to be accepted and tx to continue.

Not yet done any broadcasting of commitments; seems like something that could be added as an option, but maybe better not to in the beginning.

Not yet added a value lower limit or age lower limit to the commitment being valid; the former is a trivial change, the latter needs some tweak to utxo_set querying I think.

The amendments as listed have made the !auth message a bit longer. I think we could improve this (and the !tx message too) by shifting everything to binary, but that can be done later.

Fairly extensive tests in `test/test_podle.py`.

I'll update the 0.2.0 protocol gist and continue, let me know what you think.

**Comment by AdamISZ** [comment]
Fairly large edits to https://gist.github.com/AdamISZ/baf93ce2589854a7992383b3c69fae13 in light of the above. It's more detailed/correct now, not really more readable though :)

**Comment by chris-belcher** [comment]
Great work.
I've been following it from the commits and comparing the branches (https://github.com/JoinMarket-Org/joinmarket/compare/develop...0.2.0)

**Comment by AdamISZ** [comment]
copy from freenode:

"waxwing> for those interested in testing the new 0.2.0 , it's now set up on #joinmarket-pit2-test (i added the '2' for channel ownership), two yg bots currently running, -basic only (bugfix needed for others that i'll do next), on testnet of course, use latest on 0.2.0 branch which will change regulary so i'll be restarting the ygs regularly.
<waxwing> my 2 are currently on rizon and freenode. any others are OK, channel topic should be set but it's not a big deal."

**Comment by AdamISZ** [comment]
f0ec87e58988efb47fa06d3ac08509ee49b3e1f0 and 766ada4cffe9debe416a8c85006803e231e4dc93 are bugfixes to messaging. I've done 3 transactions in the test pit so far, will continue.

**Comment by AdamISZ** [comment]
49f66786f3db0083db20d64f8709c91e387c2946 :  prevent sig replay cross-mc with mc identifier in signed msg. This *seems* sufficient; sig replay is technically possible if you lose control of the nick on one channel and get spoofed there, but given the anti-MITM e2e encryption, it seems there is no realistic window to get non-public information. If anyone has a different idea about how to prevent replay, or a reason why this is not sufficient, please chime in.

**Comment by AdamISZ** [comment]
I thought about #90 again and came to much the same conclusion as before, although of course things are a little different now in the protocol. Since the MITM has to impersonate both sides, the goal of preventing privacy leak is achieved with just the maker side. Here's what I think we can do to change the !auth:
 
```
4: T*: !auth IUP B(tencpubkey) IU O (NS)
```

```
4: T*: !auth  O (NS)
```

Reminder of meaning of O:

Commitment data for default PoDLE construct:

**O**: commitment opening serialization, consists of the following fields separated by a field separator (default '|'):

 **U**: utxo used for PoDLE (not necessarily from wallet)
 
 **P**: pubkey used for PoDLE (not necessarily from wallet)
 
 **P2**: shifted base point pubkey for PoDLE

 **s, e**: schnorr sig values for PoDLE


A MITM - which has to realistically be the server operator, since all of this happens in private messages - would really just be relegated to being another "Taker" actor in this scenario, since they can get the maker to release an `!ioauth` message with utxo information by providing a valid (unused) commitment C in the `!fill` message, but they can go no further as they would have had to provide a separate ECDH pubkey to their taker "victim" and thus the bitcoin signature in the !ioauth is not replayable forwards (does not validate against the MITM's ecdh pubkey).

So this is effectively no worse than any DoS by denying a commitment, which remember are individually one-time-only but configurable reuse number using `taker_utxo_retries` and apply different base points.

Will code this up now. Let me know any thoughts.

**Comment by AdamISZ** [comment]
2adeb30573163caaed13c458756870d11cde5741 see the above comment for details. NOTE: user-level scripts, in particular `create-unsigned-tx.py` (not sure if others?), have not been updated; sendpayment and yieldgenerator and tumbler should be unaffected. I expect a final "sweep" through user level scripts once alterations are complete can include this.

**Comment by AdamISZ** [comment]
83f27036aaa0ac99fee258ebd9b1eab37b05900e : a rudimentary tool `add-utxo.py` that imports and optionally verifies utxos to be used as commitments from other wallets. Problematic element is that it requires a privkey (WIF compressed of course). Optionally reads from file instead. I think in future it should be tweaked in a couple of ways: include the option of not having a dependency on joinmarket code, so it can be employed on cold/different machines for security reasons (but difficult as it relies on PoDLE code which uses the privkey), and in the other direction an easier tool can be made to import commitments from other *joinmarket* wallets (there the privkey stuff can be under the hood).

I've done some basic testing, it seems to work in the 3 or 4 different modes. As usual --help tells you what you need to know.

**Comment by AdamISZ** [comment]
12aeeb10a8ac3b755952fdd9946242975e6b5de8 More exotic ygs like -deluxe send very long order announcements, and in privmsg the nick signature may spill over into a second line (at least); the previous code in `send_raw` just hardcoded a check for 'relorder' or 'absorder' in the line before deciding on throttleQ or obQ, and since the second (or last) line doesn't have that string, it was sent on the normal pingQ, thus getting processed first, resulting in messages sent out of order. This fixes that and de-hardcodes the ordertype strings (reading from the global `jm_single().ordertypes` instead)

**Comment by chris-belcher** [comment]
With the way you've coded `add-utxo.py` now, the private key will appear in `.bash_history`, `ps` and similar, which can sometimes be read even by another user on the same VPS.
I'd recommend that the same thing is done as in `create-unsigned-tx.py` where the privkey is requested via `raw_input()`. (I can create a PR if you're doing other stuff, its only a small edit)

**Comment by AdamISZ** [comment]
Ah of course, I knew there was a sensible way of doing it ... I put a big warning in the help, but somehow that escaped me ... thanks, pls do it if you have the time.

**Comment by AdamISZ** [comment]
96fcba97a03183e9d2660ee8a16f4196b301090f : allows taker to choose a utxo from the rest of his wallet, only if the utxos in the transaction are no longer available for commitment (previously code only allowed to use utxos in this tx, which is far too limited). Logic of ordering is that is preferable to use the utxos about to be consumed (although not sure it's a critical distinction, but can't hurt).

Additional minor refactoring: made the `make_commitment` function a property of `Taker` as was originally intended, and hook it into `CoinjoinTx` as a callback. This way new `Taker` inheriting classes can make their own commitments, as long as they stick to the syntax described above for commit/open plus flag byte.

Created a test and seems to work fine. Note that `test_podle.py` is a mess of copy paste code and arbitrary values at the moment, I am just using it for testing out scenarios, won't put it into the build test. Anyone else can run it of course, it passes at the moment.

**Comment by AdamISZ** [comment]
595f2e35db6c50862cdc68391d1218ccee1c7112 adds age restriction (using default 5 blocks in config). Added optional confirmation number to `query_utxo_set` return dict in `BitcoinCoreInterface` and `BlockrInterface` to support. Updated the utxo commitment choice to be: 1/utxos from tx 2/utxos in wallet 3/external for priority. Added an extra test. The most notable issue I found was described in the comment [here](https://github.com/JoinMarket-Org/joinmarket/blob/595f2e35db6c50862cdc68391d1218ccee1c7112/joinmarket/taker.py#L87-L102) - if we fail to find a commitment, I think we should just shut down fully, but I am a little uncertain that that might depend on the bot type (most notably tumbler).

As briefly discussed on IRC, might be sensible to go with fairly light handed defaults for now (5 blocks). Very large age requirement might make tumbler impractical (although even that isn't clear).

Will do coin amount next, as a fraction of coinjoin amount, probably just 20% of that as a default. Should be fairly simple.

Hopefully the basic structure of the protocol is done at that point.

**Comment by AdamISZ** [comment]
d08285e4f2ec23b854a3e0cfa5f08adcaeef4828 adds the coin amount restriction, default 20% of coinjoin amount. Tests are passing, done a couple of txs in the pit OK.

**Comment by AdamISZ** [comment]
Fixed the privkey entry insecurity in `add-utxo.py` as per your correction @chris-belcher , thanks, see 8510689d87c878c6df75c80a0d59967e5f94a05c . Also added a tool I've been using a while called `ygrunner.py` in f330bc07a3539a8af0879129c51ac70cf80d71fa that lets you run some ygs on regtest and provides you a seed for a Taker bot to run separately. Just for debugging, irrelevant otherwise.

**Comment by AdamISZ** [comment]
TODOs:

Fix `create-unsigned-tx.py` now that it now longer needs an authorizing key/utxo.

Fix the test_tx_broadcast test which is now failing: I believe it's to do with a limitation i found while building the mc-collection code - iirc I couldn't support sending to random maker.

The secp256k1 thing is a big remaining issue to address, consider: (1) the current podle code requires secp256k1, it's theoretically possible to build it on the old pybitcointools code, but I don't want to. (2) the donation code requires the old code, but breaking news: my custom nonce PR was finally merged by ludbb, meaning it could be updated to secp256k1 (great - but it's another chunk of work and especially testing). Side note another PR has been merged into that repo, that looks like it will make installation a bit smoother, but it'll need investigation. And the last thing to mention is that removing a lot of legacy code by removing non-secp256k1 is a good motivation, but again it's a huge chunk of work, so I'm ambivalent about it. I want to leave things as they are given all this (*for now*), but tell people that they simply have to install the binding.

**Comment by AdamISZ** [comment]
Re: the remaining points in the OP that haven't been addressed. I see #83 and #120 as obviously important, but difficult, and arguably not as urgent. I would like to go ahead without them (somewhat in line with the discussions we had on IRC where we said let's just focus on the really important changes).

Another point is, would it make sense to update relorder and absorder to new names, so as not to clash with the old ones, thus old bots would still be able to run in the same pit, albeit with the JM alert?

**Comment by AdamISZ** [comment]
The commit in [this] comment was dumb; of course, the server and port do *not* identify the server, since different people connect on different server names (including onions) and ports. So, fix in e63f83fe9c81c26cef89182b091998edbe0ed7a1 which is a bit clunky but I see no alternative - use a `hostid` field which uniquely identifies the server no matter how you connected. See the comment [here](https://github.com/JoinMarket-Org/joinmarket/blob/e63f83fe9c81c26cef89182b091998edbe0ed7a1/joinmarket/configure.py#L112-L115) to explain how it's set. Tests out OK. If anyone knows a more elegant solution to preventing replay attacks cross-message channel do say, thanks.

**Comment by AdamISZ** [comment]
8f50282a995501449f2a9404a3928c30c5467e85 removes the legacy signature conversions.

**Comment by chris-belcher** [comment]
Yes it makes sense to rename absorder and relorder. I suggest "absoffer" and "reloffer" because in hindsight the name "order" is confusing.

**Comment by chris-belcher** [comment]
I hope one day we figure out an alternative to the age restriction because it's not always great for user experience; often people want to receive coins and then spend them right away.

**Comment by AdamISZ** [comment]
Re: bad user exp, yeah, but it won't be a problem for those with "fuller" wallets. I guess a 1hr delay is quite annoying for a completely new user but given all the other steps you have to go through to get started, I feel like it's not a dealbreaker. At least it doesn't cost or lock up any money. And a heavy user may choose to create external commitments as a backup to avoid problems.

Re: ordernames, I like the suggestion, a little worried that it confuses with all the other stuff in the code; but, let's go with that, we could change the use of the word "order" later in the code and logs without breaking anything.

**Comment by AdamISZ** [comment]
f1524951aa4b48b1e3e01444a759922b234a22d1 PROTOCOL BREAK (Let's write this from now on for any commits that change the proto - hopefully there should be very few more, just feature adds/changes)

All hardcoded relorder/absorder changed to reloffer/absoffer. This could be encapsulated better to remove the hardcoded instances at the top level; but that can be done later easily. Note it *does not* change the use of code variables called 'order' anywhere. Tests OK. I deliberately checked the different ygs in the tests too.

**Comment by AdamISZ** [comment]
fe9f6d53d65356c6f6eeec79fb2933d0c948489a prints out detailed debug info in a file `commitments_debug.txt` in case of failed commitment generation. It shows (a) the list of utxos which were rejected due to too many retries, (2) the list of utxos which were rejected due to being too new and (3) the same for too small. It gives brief advice for each case, and at the end prints out the full set of utxos in the wallet also. The user is pointed to this file in the terminal output.

This is a bit crude (just a single re-used static file), but is the least that's needed to give a user some guidance. A more sophisticated approach would be better; note that it's written to a file as it's too much stuff for a terminal output (which is already super noisy). This is the kind of thing where a GUI really helps.

I used the `ygrunner.py` tool to do a bunch of manual cases to check it produces sensible output.

Here is an example of `commitments_debug.txt` : 

```
THIS IS A TEMPORARY FILE FOR DEBUGGING; IT CAN BE SAFELY DELETED ANY TIME.
***
1: Utxos that passed age and size limits, but have been used too many times (see taker_utxo_retries in the config):
None
2: Utxos that have less than 9 confirmations:
28d46cacb295d493a62c5bdc4486febd31ed9704cbeb77e26b5ef579a0251091:1
0a143a3a228a8cba93e492a17194aa04c5d3633cb91fe4e03bdfe0a2c730f268:0
ab6c495fbf4b8d2105a7aa1afb462ec31e70469725e62e768914849a1b1f6761:1
db71144041842e5625ae68c268d8ece2079d1b84593af06eddb37fead8f04865:0
e3174ba8c8bb88d67e737095c79cd8f21ee8807be9ea6dc1082f31a8f960b042:1
3: Utxos that were not at least 20% of the size of the coinjoin amount 199164661
None
***
Utxos that appeared in item 1 cannot be used again.
Utxos only in item 2 can be used by waiting for more confirmations, (set by the value of taker_utxo_age).
Utxos only in item 3 are not big enough for this coinjoin transaction, set by the value of taker_utxo_amtpercent.
If you cannot source a utxo from your wallet according to these rules, use the tool add-utxo.py to source a utxo external to your joinmarket wallet. Read the help with 'python add-utxo.py --help'

You can also reset the rules in the joinmarket.cfg file, but this is generally inadvisable.
***
For reference, here are the utxos in your wallet:

{u'28d46cacb295d493a62c5bdc4486febd31ed9704cbeb77e26b5ef579a0251091:1': {'value': 100000000, 'address': u'n19n1b9LEqUp2T8FJS9t7NEQXiKhwwvbNi'}, u'0a143a3a228a8cba93e492a17194aa04c5d3633cb91fe4e03bdfe0a2c730f268:0': {'value': 100000000, 'address': u'miTgCc9DQjRaiuyUN3JTomjXmTGMt9DaA5'}, u'ab6c495fbf4b8d2105a7aa1afb462ec31e70469725e62e768914849a1b1f6761:1': {'value': 100000000, 'address': u'mhZJMVoZAv7BzJqBMpfodUxQEtPrSi3fLj'}, u'db71144041842e5625ae68c268d8ece2079d1b84593af06eddb37fead8f04865:0': {'value': 100000000, 'address': u'mxtRWHA87F4RRs7CtJbXBgoCcPuk5eWbGQ'}, u'e3174ba8c8bb88d67e737095c79cd8f21ee8807be9ea6dc1082f31a8f960b042:1': {'value': 100000000, 'address': u'mmy54TphEoHmrruJXP2XkWQVSwo6ieYbxr'}}
```

While doing this I noticed something very important that I hadn't noticed about `sendpayment`, which happens to affect the commitment creation algo. `sendpayment` previously constructed a `Wallet` object only up to the used spending mixdepth; but to get a commitment from the whole wallet need an object containing all mixdepths. For this reason, added an option `-a` to specify the actual max mixdepth of the wallet so all utxos get pulled in for commitment generation. People using the default 5 won't be affected of course.

Tests passing.

**Comment by AdamISZ** [comment]
c5860c64b82378abc68c7a739d7c73d8f51f40d2 nkuttler noticed when I removed the legacy sig conversion - bot can crash on an invalid signature during verify (asserts in the Python binding code), so wrapped the parsing in a try block; only for verify, not sign of course.

**Comment by AdamISZ** [comment]
db8a2aeb91f336ae47e32a56758a443596158502 PROTOCOL BREAK - 
from @chris-belcher to use the announced `network` field from the IRC server to specify the `hostid` variable (obviously it's intended this could also be set by other non-IRC networks, that can be done later). Doing some live testing in the testpit today. Did one tx OK with my own bots(not others due to protocol break), but noticed that the network name for Freenode doesn't seem to be getting picked up. Will look at it later.

**Comment by chris-belcher** [comment]
I missed the : thing in testing, I should've tried it on many different IRC networks in hindsight. Will fix soon.

**Comment by AdamISZ** [comment]
np, and thanks. btw feel free to push the fix directly to the branch.

**Comment by AdamISZ** [comment]
a255d5d993ad16dcc11fe6a8da268bf991a1ae3f allows Makers to broadcast used commitments to the pit with a command `!hp2 <hash>`. Makers can choose to ignore or not these broadcasts with the config variable `accept_commitment_broadcasts` in `POLICY`. Note this is *not* a protocol break.

An important subtlety is when the broadcast occurs: in a typical scenario `!fill`, containing the commitment, is sent out to N makers at the same time; if they broadcast *immediately*, this could lead to other makers taking part in the transaction erroneously adding it to their blacklist before processing it themselves; therefore, the value is broadcast immediately if acceptance fails (i.e. if it's already in *my* blacklist), or otherwise broadcast after sending `ioauth` (i.e. after all checks have passed).

Important to remember in this that each individual hash value/commitment is indeed intended to be used only once globally, that's why we have "retries" that allows multiple (default 3) for the same utxo. I ummed and ahhed about this broadcast element for a while, but the more concretely I think about it the more I'm convinced it should be used to get the real rate-limiting effect. So, currently have the `accept_commitment_broadcasts = 1` in the default config, i.e. switched on. Switching it off is always an option, there is no issue with agreement/consensus here.

Tests passing.

**Comment by AdamISZ** [comment]
891cfe9982d789148caaad2515c9b10478a88aff is designed to reduce the effect whereby broadcasting commitments on the public pit channel unambiguously marks the maker as a participant in a transaction, making it easier for a spy to focus their queries on certain maker bots. This is accomplished with a new method `Maker.transfer_commitment` which chooses another counterparty with active orders (for this purpose, as suggested by @chris-belcher , the `Maker` class now inherits from `OrderbookWatch` rather than `CoinjoinerPeer` - a change which I don't believe has any undesirable side effects, afaict) and sends the commitment to them with `hp2` command name as before, over private message. The recipient, having a `on_commitment_transferred` callback, simply broadcasts it with a public message to the pit. Thus the origin of the commitment broadcast is obfuscated.

Worth noting the possible imperfections - the receiving party may simply choose not to broadcast; the receiving party may be acting on behalf of a snooper, or be a snooper. But, this seems unambiguously better than direct public broadcast, since failure to broadcast is not in anyway critical (this is especially true since there will usually be ~ 3-5 different makers broadcasting the commitment at once).

Note that none of this changes the existing `accept_commitment_broadcasts` setting either way; that controls whether a maker responds to broadcasted commitments (i.e. adds them to their blacklist); that code is only triggered by the public message.

Not a protocol break.

**Comment by AdamISZ** [comment]
d50ae7c053dfa3043eaf1d979d51f075e6c83c6d addresses a bug: if a bot cycles nick on one channel it doesn't on the other, leaving the message channel layer confused. Just changed get_irc_nick so that the fixed length (without underscores) nick is picked up.

**Comment by chris-belcher** [comment]
Fixed bug in IRC network name finding https://github.com/JoinMarket-Org/joinmarket/commit/66fdcce19785c1207e22c86f1b153191ceec7360

Protocol break because Freenode is correctly found now

**Comment by AdamISZ** [comment]
53e30d74e0f83bf3e02b41ab1b7f54d4a556a5b0 seems to be the most logical way to deal with nicks changing, after discussion with @chris-belcher : force the nick/username change on all channels if it happens on one. I'd like to get a solid test of this, but it'll take some considerable work, and it's a very unusual condition, so will start running it in the test pit straight away. Review would be appreciated.

**Comment by AdamISZ** [comment]
I'm thinking of making a yg-algo that has the best privacy effect, at the moment all I have is: offer announce is max from maxmixdepth (so one price only, not like current *mixdepth), and requests are responded to with utxos from the appropriate mixdepth depending on requested amount, then (here a big difference to earlier perhaps): the cjout goes to some *other* mixdepth than maxmixdepth, to avoid reannouncement. Two obvious limitations are: this obviously will not *always* avoid reannouncement (presumably impossible!), OK that's fine, but also: coins tend to de-concentrate. Thoughts?

**Comment by AdamISZ** [comment]
30d69a099db27d136c828282abe1adcce3e0ba22 does 2 things: (1) a `yieldgenerator.py` module within `joinmarket`, which provides a base class for YGs to extend (with abstract methods for `create_my_orders` etc). This wasn't too hard at all really, although it's arguable whether `ygmain` with the option parsing should go in there. Anyway, it's fine for now, and avoids a chunk of code duplication; it can be refined later.

Second, re the previous comment, created `yg-pe.py` meaning "privacy enhancing" as a new yield generator. It only implements the simplest (and I *think* the soundest) of the ideas listed above: minimize re-announcement by choosing to source utxos not from the largest mixdepth, where possible and sending cjouts to, again, not the largest (in fact the smallest) mixdepth, for the same reason. Review would be appreciated; it passes regtest and I've done several runs to check it; it seems OK.

**Comment by AdamISZ** [comment]
cdcbc67d32e3162bad943b92c1aa8a48e6c0451e removes the legacy pybitcointools entirely. Some notes:

1. I've retried the secp256k1 install on fresh Ubuntu1404 and Win7 machines. It works, but there are some points of note. First, I believe that the instructions on package dependencies [here](https://github.com/ludbb/secp256k1-py/blob/master/README.md) are correct except technically missing two python dependencies: `python-dev` and maybe `python-pip` (obviously you need pip somehow). So the full dependency list seems to be:  `python-dev python-pip build-essential automake pkg-config libtool libffi-dev libgmp-dev` from a standing start. This is assuming you don't already have a Bitcoin Core installed (and visible) such that libsecp256k1 alread exists, of course. Another note is that the latest merged PR to `secp256k1-py` seems to create wheels (meaning binary distributions) automatically for new releases for Linux and OSX. And, although it refers to this in the README, there appear to be no wheels as of yet at [pypi](https://pypi.python.org/pypi/secp256k1), still only the source tarball. I'll hit up Ludvik on this at some point, although he's not often around. Once that is in place it may make life considerably easier for *some* users (although many won't want that trust model, perhaps).

With respect to Windows, the existing instructions [here](https://github.com/JoinMarket-Org/joinmarket/wiki/Installing-JoinMarket-on-Windows) seem to work exactly as before, albeit it's an ugly business (installing MinGW to get 1 DLL file).

With respect to OSX, I have no idea, but the dependencies are listed, one assumes they probably work.

2. My [custom nonce PR](https://github.com/ludbb/secp256k1-py/pull/11) was merged, and hence the donation code now uses it. I refactored the donation code, putting the address generation in the `configure` module, and tested it in `test_donations`, which is now a better test because it uses the same code, and checks that the funds are accessible using the standard address regeneration from nonce*G in the tx input signature. I also added a `donate` flag to the `test_regtest` suite, which I'm not including in the build, but ran manually a few times.

There are also a few redactions from the tests due to the removal of the old code, but that's a very minor matter; the tests were mostly already ignoring the legacy code.

I've attempted to reach out to users about this change via IRC, [reddit](https://www.reddit.com/r/joinmarket/comments/4x7jqp/who_is_not_using_the_secp256k1_binding/), bitcointalk. Not much response. I hope people can appreciate how hard it is to maintain two different Bitcoin interfaces.

**Comment by AdamISZ** [comment]
Complete tumbler test (unedited) run ran OK, so that's a good sign, although it isn't really a test of how commitments can gum things up right now since it starts with a "big" wallet.

**Comment by AdamISZ** [comment]
59443a2352bcbfd576b5f07c8ce5893917c1d656 does the following:

1. Change `create-unsigned-tx.py` (and a couple of minor changes to `taker.py`) to support the new protocol. It no longer needs an `auth_utxo` provided or private key entered; however, something similar is required: a user would have to provide external utxo commitment(s) via the `add-utxo.py` tool.

Implementing this and testing it threw up a few issues:

2. Add a `showutxos` method to `wallet-tool.py` for convenience (my use case was to use external commitments from other joinmarket wallets, but there may be other situations of course where it's useful).

3. Makers crashed when I tried to send spent utxos as unspent (commitments), added a check to make sure the gettxout return val is parsed correctly.

4. There was an insidious bug in the json parsing in the podle module that unexpectedly deleted the list of used commitments; fixed.

**Comment by AdamISZ** [comment]
c6c67bab0ebb8866a50942212e75e1f597a46fdc is trying to streamline processes for adding external commitments, or extending the utxos in a wallet, which will be important for anyone trying to act as a Taker (especially tumbler) if they're in a "from scratch" scenario - i.e. fund a new wallet and immediately run. There is a new `cmttools` directory in which the `commitments.json` will now be stored. The contents of that file are now readable (although you won't generally do that) using indent=4 option to json.dumps.

So your problem is to make sure you have enough commitments to do the job. If you're doing 1 sendpayment your life is much easier; you could just wait 5 blocks before starting, and/or you can fund with multiple utxos instead of 1. You can also add a bunch of external utxo commitments for longer term use.

With tumbler your job is harder: even funding with multiple utxos and waiting 5 blocks might not be enough, since they are liable to get consumed during the run. You can make the wait times long enough that they safely cover 5 blocks, but that's long. If you're starting with a fresh wallet you may well need to follow the "external commitments add" workflows described here.

So there's various workflows for this: (a) you want to add external commitments, (a1) you have an existing joinmarket wallet (a2) you only have other wallets (like Core, Electrum), (b) you want to make a bunch of utxos in your wallet instead of only 1, (b1) you have an existing joinmarket wallet (b2) you only have other wallets.

(a) `add-utxos.py`
(a1)
An easy two-step process to grab all the utxos from another wallet (like one used for a yieldgenerator) and use them for commitments as spending in a new wallet (of course, it will generally be far easier to just use the same wallet, then no messing around like this).

Run `python wallet-tool.py -p -u walletname showutxos` -> output is walletname.utxos (e.g. wallet.json.utxos) in the wallets/ directory containing utxos and corresponding private keys. Then `cd cmttools ; python add-utxos.py -R ../wallets/walletname.utxos` (there is an option to verify the utxos as valid unspent first using the -o option). You should then see these added into commitments.json in the "external" field. Then delete the file with the private keys: `rm wallets/walletname.utxos`.

(a2)
More realistic case: you have utxos say in Electrum. You'd need to gather the utxos and their corresponding private keys and enter them into a csv file one by one (each line format txid:N, wif-compressed-privkey). Then do `python add-utxos.py -r csvfile` to achieve the same result as before. And again delete the file with the private keys...

One can also enter single utxos with `python add-utxos.py txid:N` and enter wif-comp-key on command line prompt (as already discussed for the first version of `add-utxos.py`).

The script has other minor options like delete and verify utxos too.

(b1) Second tool: `cmttools/sendtomany.py`
To produce several utxos from one: `python sendtomany.py utxo destaddr1 destaddr2 ...`. This does as you'd expect; send the coins with equal amount after fees into each of several destinations (intended to be different "external" addresses in a joinmarket wallet). 

(b2) This send-to-many feature exists in other wallets like Electrum, Core (I think?). This is obviously better where possible, you can just send 1 utxo to 5-10 new addresses in your joinmarket wallet, instead of only to 1, without doing anything difficult.

The bottom line as I see it: we want to use (b2) and in any case, not (a), because it doesn't require any exotic signing and therefore private key manipulation. But I doubt one can really run tumbler this way (even spreading deposit across multiple mixdepths can't work since it breaks the privacy model), and so, since we want to keep instructions simpler for users, I'm more inclined to tell them to use (a) as a mostly one-off job when they start: give yourself say 5-10 utxos (so 15-30 "free tries") from another wallet, using (a2) probably. But this does mean dicking around with private keys. It could also mean some extra privacy loss to makers since they read the utxo that was used on commitment opening.

Perhaps the only way to cut the Gordian knot here is to require tumblers to run with a wait > N blocks, and perhaps 5 is too aggressive?

**Comment by AdamISZ** [comment]
I've been doing a lot of testing to see how this looks in practice. A couple of things:

* I think the best way to deal with the above problem is simply to add a config field that a tumbler can set that means "if a commitment can't be found, wait a few minutes and try again". This means the tumbler will run at a rate set by the user, until it runs out of commitment utxos, then just waits ~5 confirmations before resuming. It's a simple enough solution and doesn't require any interaction, but will make a tumbler run longer than it was before, in most scenarios.

* The tools I made for adding commitments involving creating a file with private keys, that's making it much too easy for a user to have private keys lying around, a bit ridiculous. I'll make amendments so those files are encrypted with the same key as the wallet.

**Comment by AdamISZ** [comment]
82f09daa5249778cfa414c48d096c8d9309b5d3f couple of bugfixes and: add-utxo reads wallet directly rather than creating a file with private keys as above (not secure + inconvenient).

**Comment by AdamISZ** [comment]
dcf93334ffcd2a51a4d6afe2e006b299dca64ff1 - update the `test_broadcast_method.py` to work with the new system (dummy commitments). Now all the test suite passes.

a4ac313a789a29feeec3dce26f00d681875ba433 was essentially a bugfix, because externally sourced commitments weren't being checked for age/size limits, now they are.

Full tumbler has been run a couple of times, done a variety of other transactions, I think this is largely stable now.

I even wrote some primitive attacker code to try to sweep up utxos from makers, although it needs a lot of work. But with only a few bots in the pit and no real activity, there's not much that can be done in terms of simulation.

Spent some time trying to get sensible measures of how many utxos the spy needs; it's nearly impossible, there are so many variables\*, but overall I get the impression that at current levels they really won't need much. Could be anything from 10-200 in an hour I think, for the current pit ("hour" is a sensible measure if we stick to ~5 block age requirement, since that's the "refresh rate" they'll need over the long term). And they have to have the right amount range, of course.

I think the deterrent factor of creating fresh utxos all the time will be limited, but a combination of people using more "guarded" yg algos that don't advertise tons of information, forcing them to use up more utxos, plus the fact that their own utxos will be very visible and easily identified, might change the dynamics. That's the short term perspective, the long term perspective, if being optimistic, is that if we can scale up, it'll get harder and harder for them.

\* Here are some I looked at; some of them aren't easy to define, let alone measure: 

* Pit Tx/hour - global coinjoin rate
* N makers - 50 today, 100 in the future?
* Utxo age reqmt - (5 blocks now)
* Utxo retries allowance - (3 now)
* Mixdepths/maker - 5 default now
* Tx btc size (mean) - (a better model would involve a distribution, perhaps power-law, but it's completely murky anyway, maybe a spy is only interested in the bigger ones)
* Maker-requests/commit - How many maker requests you can make simultaneously with the same commitment.
* Makermixdepthfindfactor - Proportion of mixdepths you need to query before finding the one with the utxo you're looking for. This factor strongly depends on how visible are the mixdepth boundaries. It might be larger than one if you don't know them.
* Maker success factor - Proportion of makers you need to query before finding the utxos you're looking for. This is lower than it might seem because this is a “dragnet” operation, trying to find all cjouts in **all** relevant transactions.

**Comment by AdamISZ** [comment]
a9ede54970c88b4c3303b56ecfe2a4f5d1df3901 - noticed that if multiple bots use the same directory (not something planned for, but in any case), blacklist file being shared means transactions getting unnecessarily blocked. This pushes back the *adding* commitments to the blacklist to the point they're actually used, i.e. on sending `io_auth`.

**Comment by AdamISZ** [comment]
Minor bugfix 980c53643a3b738a720c5ab5cc78ea76df5f313c README update (also updated/added wiki articles) a6f191e2fd2620ed8cc0ad11ef631ed28a3684a8 , which includes reference to 00301e16905f35d896a7134e1a4ca4207bebe259 which formalizes external requirements `libnacl` and `secp256k1(-transient)` for installation. Also now testing successfully on anarplex (and got conf from server owner that it's OK, at least for now) so tentatively suggesting default server set:

```
[MESSAGING]
host = irc.cyberguerrilla.org, agora.anarplex.net, irc.rizon.net
channel = joinmarket-pit, joinmarket-pit, joinmarket-pit
port = 6697, 14716, 6697
usessl = true, true, true
socks5 = false, false, false
socks5_host = localhost, localhost, localhost
socks5_port = 9150, 9150, 9150
```

with this alternative for those wishing to connect to HS:
```
[MESSAGING]
host = irc.cyberguerrilla.org, agora.anarplex.net
channel = joinmarket-pit, joinmarket-pit
port = 6667, 6667
usessl = false, false
socks5 = true, true
socks5_host = localhost, localhost
socks5_port = 9150, 9150
```

It seems like if the default uses all 3, then a maker requiring Tor only is not at a disadvantage, but, if you disagree, let me know. If a third supporting Tor+clearnet option is available, we can add/change to that.

**Comment by AdamISZ** [comment]
4146a7d6790de78b501bfd91d4958ee9f5348143 removes all yield generators except yg-basic and yg-pe ; the former inheriting from YieldGenerator in the joinmarket module so there is no longer significant code duplication.

The intention is twofold: (1) can't support multiple customised yield generators, it's impractical, but doesn't mean an attempt to stop them: I'd suggest making a separate repo in the Org to host other implementations, starting with what we already had - but note they actually have to be supported by coders. (2) emphasize that simpler, less information providing yield generators are better for privacy and for the overall system. This point is not completely uncontroversial, of course, but again: this is not an attempt to "ban" customizing, it's an attempt to limit the scope of the core project.

**Comment by adlai** [comment]
Closing this issue now, as the `s/rd/ff/` protocol upgrade (f152495) has been frozen by release. Addressing issues left pending (#120, #29/#45, #27/#423) thus requires another "protocol softfork" (eg `s/(abs|rel)offer/$0v2/`), distinct from the one midwifed in this thread. Discussion should continue in respective threads.
